src/graphics/shader-processor.js

```diff
@@ -11,14 +11,46 @@ import { BindGroupFormat, BindBufferFormat, BindTextureFormat } from './bind-gro
 /** @typedef {import('./graphics-device.js').GraphicsDevice} GraphicsDevice */
 
 // accepted keywords
-const KEYWORD = /[ \t]*(\battribute\b|\bvarying\b|\bout\b|\buniform\b)/g;
+// TODO: 'out' keyword is not in the list, as handling it is more complicated due
+// to 'out' keyword also being used to mark output only function parameters.
+const KEYWORD = /[ \t]*(\battribute\b|\bvarying\b|\buniform\b)/g;
 
 // match 'attribute' and anything else till ';'
 const KEYWORD_LINE = /(\battribute\b|\bvarying\b|\bout\b|\buniform\b)[ \t]*([^;]+)([;]+)/g;
 
 // marker for a place in the source code to be replaced by code
 const MARKER = '@@@';
 
+const precisionQualifiers = new Set(['highp', 'mediump', 'lowp']);
+
+class UniformLine {
+    constructor(line) {
+        this.line = line;
+
+        const words = this.splitToWords(line);
+        this.precision = undefined;
+        this.type = words[0];
+        this.name = words[1];
+
+        // if this is a precision qualifier
+        if (precisionQualifiers.has(this.type)) {
+            this.precision = words[0];
+            this.type = words[1];
+            this.name = words[2];
+        }
+    }
+
+    splitToWords(line) {
+        // remove any double spaces
+        line = line.replace(/\s+/g, ' ').trim();
+        return line.split(' ');
```

Comment:
You could also split it right away via RegEx, I don't know if it feels easier:
```suggestion
        return line.trim().split(/\s+/);
```